File: src/main/scala/com/qunar/spark/tungsten/api/java/JDataSets.java
package com.qunar.spark.tungsten.api.java;

import com.fasterxml.jackson.core.type.TypeReference;
import com.qunar.spark.tungsten.api.DataSet;
import com.qunar.spark.tungsten.base.SparkSessions;
import com.qunar.spark.tungsten.base.TypeConverter$;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.SparkSession;
import scala.reflect.ClassTag;
import scala.reflect.api.TypeTags;

/**
 * 生成与使用{@link org.apache.spark.sql.Dataset}的通用java入口类
 */
public class JDataSets {

    private SparkSession sparkSession = SparkSessions.getSparkSession();

    private <T> DataSet<T> createFromDataset(Dataset<T> dataset) {
        return new DataSet<>(dataset, ScalaConverterDriver.getTypeTagOfT());
    }



    /**
     * scala-java转换的驱动封装,隐藏复杂的scala api调用
     */
    static final class ScalaConverterDriver {

        /**
         * 借用Jackson的{@link TypeReference}获得泛型T所对应的实际类型
         */
        @SuppressWarnings("unchecked")
        static <T> Class<T> getClassOfT() {
            return (Class<T>) new TypeReference<T>() {
            }.getType();
        }

        /**
         * 由{@link Class<T>}转换为{@link ClassTag<T>}
         */
        static <T> ClassTag<T> getClassTagOfT() {
            return scala.reflect.ClassTag$.MODULE$.apply(getClassOfT());
        }

        /**
         * 由{@link Class<T>}转换为{@link TypeTags.TypeTag<T>}
         */
        static <T> TypeTags.TypeTag<T> getTypeTagOfT() {
            return TypeConverter$.MODULE$.classToTypeTag();
        }

    }

}

File: src/main/scala/com/qunar/spark/tungsten/api/java/JDataSets.scala
package com.qunar.spark.tungsten.api.java

import com.hadoop.mapreduce.LzoTextInputFormat
import com.qunar.spark.base.json.JsonMapper
import com.qunar.spark.tungsten.base.CommonEncoders._
import com.qunar.spark.tungsten.base.{CommonEncoders, SparkSessions, TypeConverter}
import org.apache.hadoop.io.{LongWritable, Text}
import org.apache.spark.api.java.function.MapFunction
import org.apache.spark.sql.Dataset
import scala.language.implicitConversions

/**
  * 生成与使用[[JDataSet]]的通用java入口类
  */
object JDataSets extends Serializable {

  private val sparkSession = SparkSessions.getSparkSession

  /**
    * 给定一个[[Dataset]],创建对应的[[JDataSet]]
    */
  private[tungsten] def createFromDataset[T](dataset: Dataset[T]): JDataSet[T] = {
    new JDataSet[T](dataset)
  }

  /**
    * 读出原始内容(text或gzip)并转成[[JDataSet[String] ]]
    */
  def readTextFromHdfs(path: String): JDataSet[String] = {
    val dataset = sparkSession.read
      .text(path)
      .as[String]

    createFromDataset(dataset)
  }

  /**
    * 读出原始内容(text或gzip)并转换成[[JDataSet[T] ]]
    */
  def readTextFromHdfs[T](path: String, convert: MapFunction[String, T]): JDataSet[T] = {
    val dataset = sparkSession.read
      .text(path)
      .as[String]
      .map(convert, CommonEncoders.encoderForSingle(TypeConverter.classToTypeTag[T]))

    createFromDataset(dataset)
  }

  /**
    * 读出LZO压缩文件并转换成[[JDataSet[String] ]]
    */
  def readLzoFromHdfs(path: String): JDataSet[String] = {
    val dataset = sparkSession.sparkContext
      .newAPIHadoopFile[LongWritable, Text, LzoTextInputFormat](path)
      .values
      .map(_.toString)
      .toDS

    createFromDataset(dataset)
  }

  /**
    * 读出LZO压缩文件并转换成[[JDataSet[T] ]]
    */
  def readLzoFromHdfs[T](path: String, convert: MapFunction[String, T]): JDataSet[T] = {
    val dataset = sparkSession.sparkContext
      .newAPIHadoopFile[LongWritable, Text, LzoTextInputFormat](path)
      .values
      .map(_.toString)
      .toDS
      .map(convert, CommonEncoders.encoderForSingle(TypeConverter.classToTypeTag[T]))

    createFromDataset(dataset)
  }

  /**
    * 将[[JDataSet[T] ]]写入hdfs
    */
  def writeToHdfs[T](content: JDataSet[T], path: String): Unit = {
    content.getInnerDataset.map(record => JsonMapper.writeValueAsString(record)).write.save(path)
  }

}
